### DISCUSSION TITLE
Whats wrong with unit for Total Energy? Or is it me?

### DISCUSSION THREAD

**Discussion by abuesen** [Hyundai-Kia-Connect/kia_uvo#817] [Q&A]
I'm trying to understand "_Total Energy Consumption_" and "_Total Energy Regeneration_". We own an HY-Kona long range bought in 2021. 

_Odometer_: 53.206,8 km. 
_Total Energy Consumption_: 1.136.874 Wh
_Total Energy Regeneration_: 429.913 Wh
Car battery capacity is 64kWh.

**First** in my opinion the _Total Energy_ has the wrong measurement-unit. It should be 'Deka-Wh', what is very unusal.

If I calculate kwh/100km it is 1136kWh / 53206km * 100 = 2,135 kWh per 100km.
That would be nice but it is much less than possible. The 21,35 kWh per 100km seems a bit high but it is plausible.

If I calculate the count of loading-cycles it is 1136 kWh / 64 kWh = 17 cycles. Not plausible.

So the _Total Energy Consumption_ must be  11.368.740 Wh. There is one digit missing at the entitiy.
Am I wrong? Do I miss something?

**Second** I try to understand how the values fit together. I track many loading-cycles, it is always below 21kWh per 100km I need from grid. Different loading points. And I know the consumption the Kona tells, it is between 14 kWh (summer) and 19 kWh (winter) per 100km.

So one see in the above calculation: The _Total Energy Consumption_ divided by km is to hight.

If I calculate _Total Energy Consumption_ - _Total Energy Regeneration_ divided by km it is far to low.
1136874  deWh -  429913 deWh / 100 / 53206km * 100 = 13,288 kWh per 100km

So what do the values represent?

**Comment by MichT368** [comment]
Indeed, there must be a mistake inside punctuation.

And like you: I never really understood, whether the energy_regeneration adds to energy_consumption or not.

If you take the results of my car (already adjusted, due to your suggestion), you have the following totals:
```
Odometer: 46769 km
total_energy_consumption (adjusted):  6537,85 kWh
total_energy_regeneration (adjusted):  1784,63 kWh
```
If I add those two, assumed, that this is the correct  point of view, I get
```
8322,48 kWh
```
which then could be named as "new_total_energy_consumption".

The longtime calculation of consumption/100km looks as follows:
```
8322,48 kWh / 46769 km *100 km ≈ 17,8 kWh/100km
```
This seems to be a reasonable result.

**Reply by abuesen** [reply]
That's really reasonable. 
You sum up the values because this assumed meaning:

`total_energy_consumption`: Energy got from grid
`total_energy_regeneration`: Energy got from rekuperation.

I assumed this meaning
`total_energy_consumption`: Energy consumed by engine (and supplement)
`total_energy_regeneration`: Energy got from rekuperation.
`total_energy_consumption - total_energy_regeneration`: Energy got from grid.

```
11136,87 kWh + 4299,13 kWh = 15436 kWh
11136,87 kWh - 4299,13 kWh = 6837,74 kWh

11136,87 kWh / 53.206 km * 100 ≈ 20,93 kWh/100km
15436 kWh / 53.206 km * 100 ≈ 29,01 kWh/100km
6837,74 kWh / 53.206 km * 100 ≈ 12,85 kWh/100km
```

**Comment by MichT368** [comment]
Based on my assuming, the following HA Template Sensor would reflect these arguments:

```
sensor:
  - platform: template
    sensors:
      ev6_lifetime_average_energy_consumption:
        friendly_name: "EV6 Lifetime Average Energy Consumption"
        unit_of_measurement: "kWh/100km"
        value_template: >
          {% set odometer = states('sensor.ev6_odometer') | float %}
          {% set energy = states('sensor.ev6_total_energy_consumption') | float %}
          {% set regeneration = states('sensor.ev6_total_energy_regeneration') | float %}
          {% set result = (energy + regeneration) / odometer %}
          {{ result | round(1) }}
        icon_template: "mdi:ev-station"
```
No warranties, Your own risk!

**Comment by abuesen** [comment]
Ok, that is very strange. 
Todays value:

```
Energy Consumption: 1060342 Wh
Energy Regeneration: 410238 Wh
```
Both values decrease, as I found not for the first time.
Maybe there is nothing wrong, maybe this is no real "total" and only the last 100 rides count or similar.
Very strange.


![image](https://github.com/Hyundai-Kia-Connect/kia_uvo/assets/136979340/888bc0dd-dd50-4e75-b91e-a27de265a101)

**Reply by FlorianWilk** [reply]
Propably the same issue as mine: https://github.com/Hyundai-Kia-Connect/kia_uvo/issues/765

Do these decreases happen around the same time? Looking at your screenshot, it seems to be around 22:xx, right?

**Reply by abuesen** [reply]
Yes, confirmed. Both values decrease at about 22h CET. Not each day but many days.
I'll read the other Issue. 

What about the values themselves at your side? Are the punctuation correct and what about the plausability?

**Reply by FlorianWilk** [reply]
tbh not sure. I noticed the decreases and was like "this seems to be wrong", before digging deeper into the plausability of the values. I could not find a solution by now, so i decided to measure only "load from grid" using a shelly device in the fuse box of my garage. 
A slightly different usecase, but at the end i am interested in "how much money do i pay for loading the car". And the Shelly Devices seems to be much more reliable than the values from the Hyundai-API

**Comment by abuesen** [comment]
Found another project who says, values at bluelink are not reliable.
https://github.com/ZuinigeRijder/hyundai_kia_connect_monitor

But they have more entities listeted. Are they availiable at homeassistant, too?
Is it the Kona doesen't offer this values?
For Example `Engine Consumption` and `Climate Consumption`

https://github.com/ZuinigeRijder/hyundai_kia_connect_monitor?tab=readme-ov-file#python-dailystatspy

**Reply by cdnninja** [reply]
That project uses our api as the data source.  So either 

1. It isn't exposed in home assistant today but could be if an easy way to represent it. 
2. Your car doesn't have the data in the api. This can be caused by either the car not offering it in native app or us needing to adjust the api library to broaden support.

**Reply by abuesen** [reply]
> That project uses our api as the data source

Ah, it becomes full circle :-)  
I didn't install the tool, just found it and try to understand the values in HA.

> 2. Your car doesn't have the data in the api. This can be caused by either the car not offering it in native app or us needing to adjust the api library to broaden support.

In the native app I don't see any "total" but the energy consumption last week divided by _Driving_, _Electronic devices_, _Climate control_ and _Battery care_ The _Entergy consumption_ is given as "total" for one week only.

So I don't know if the values are present but not shown in HA.

**Reply by Bubo08** [reply]
You can check in HA the field Data. There you see all information which is delivered and perhaps not shown:
![image](https://github.com/Hyundai-Kia-Connect/kia_uvo/assets/12473381/004bb6cd-a463-4a9d-948c-1944c196293c)

**Reply by abuesen** [reply]
Got it. There is no energy at all. At least the values we talk about and that are seen already in HA should appear, but the do not.

```
vehicleLocation:
coord:
lat: 5x.xxxx36
lon: 7.xxxx69
alt: 310
type: 0
head: 54
speed:
value: 0
unit: 0
accuracy:
hdop: 0
pdop: 0
time: '20240205145930'
vehicleStatus:
airCtrlOn: false
engine: false
doorLock: false
doorOpen:
frontLeft: 0
frontRight: 0
backLeft: 0
backRight: 0
trunkOpen: false
airTemp:
value: 02H
unit: 0
hvacTempType: 1
defrost: false
acc: false
evStatus:
batteryCharge: true
batteryStatus: 74
batteryPlugin: 2
remainTime2:
etc1:
value: 76
unit: 1
etc2:
value: 570
unit: 1
etc3:
value: 160
unit: 1
atc:
value: 225
unit: 1
drvDistance:
- rangeByFuel:
evModeRange:
value: 278
unit: 1
totalAvailableRange:
value: 278
unit: 1
type: 2
reservChargeInfos:
reservChargeInfo:
reservChargeInfoDetail:
reservInfo:
day:
- 4
time:
time: '0640'
timeSection: 0
reservChargeSet: false
reservFatcSet:
defrost: false
airTemp:
value: 10H
unit: 0
hvacTempType: 1
airCtrl: 0
heating1: 0
offpeakPowerInfo:
offPeakPowerTime1:
starttime:
time: '1200'
timeSection: 0
endtime:
time: '1200'
timeSection: 0
offPeakPowerFlag: 0
reserveChargeInfo2:
reservChargeInfoDetail:
reservInfo:
day:
- 9
time:
time: '1200'
timeSection: 0
reservChargeSet: false
reservFatcSet:
defrost: false
airTemp:
value: 10H
unit: 0
hvacTempType: 1
airCtrl: 0
heating1: 0
reservFlag: 0
ect:
start:
day: 9
time:
time: '1200'
timeSection: 0
end:
day: 9
time:
time: '1200'
timeSection: 0
targetSOClist:
- targetSOClevel: 100
dte:
rangeByFuel:
evModeRange:
value: 381
unit: 1
totalAvailableRange:
value: 381
unit: 1
type: 2
plugType: 0
- targetSOClevel: 100
dte:
rangeByFuel:
evModeRange:
value: 381
unit: 1
totalAvailableRange:
value: 381
unit: 1
type: 2
plugType: 1
ign3: true
hoodOpen: false
transCond: true
steerWheelHeat: 0
sideBackWindowHeat: 0
tirePressureLamp:
tirePressureLampAll: 0
tirePressureLampFL: 0
tirePressureLampFR: 0
tirePressureLampRL: 0
tirePressureLampRR: 0
battery:
batSoc: 79
batState: 0
lampWireStatus:
stopLamp:
leftLamp: false
rightLamp: false
headLamp:
headLampStatus: false
leftLowLamp: false
rightLowLamp: false
leftHighLamp: false
rightHighLamp: false
leftBifuncLamp: false
rightBifuncLamp: false
turnSignalLamp:
leftFrontLamp: false
rightFrontLamp: false
leftRearLamp: false
rightRearLamp: false
smartKeyBatteryWarning: false
breakOilStatus: false
sleepModeCheck: false
time: '20240205153616'
remoteWaitingTimeAlert:
remoteControlAvailable: 1
remoteControlWaitingTime: 168
elapsedTime: '08:40:33'
systemCutOffAlert: 0
tailLampStatus: 0
hazardStatus: 0
odometer:
value: 53257.7
unit: 1
Vehicle name
KONA
```
